File: R/core/download_nse_historical_data.R
#!/usr/bin/env Rscript

# =============================================================================
# Download & Build Historical NSE Securities File: data/nse_sec_full_data.csv
# =============================================================================
# - Pulls daily bhavcopy-style equity data from NSE website for a date range
# - Appends/merges into a single, deduplicated historical file
# - Output is compatible with:
#   - load_latest_data.R (expects data/nse_sec_full_data.csv)
#   - fixed_nse_universe_analysis.py
# =============================================================================

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
  library(lubridate)
  library(httr)
})

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

get_project_root <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("^--file=", "", file_arg[1]))))
  }
  if (!is.null(sys.frames()[[1]]$ofile)) {
    return(dirname(normalizePath(sys.frames()[[1]]$ofile)))
  }
  getwd()
}

PROJECT_ROOT <- get_project_root()
setwd(PROJECT_ROOT)

DATA_DIR <- file.path(PROJECT_ROOT, "data")
if (!dir.exists(DATA_DIR)) dir.create(DATA_DIR, recursive = TRUE)

HIST_FILE <- file.path(DATA_DIR, "nse_sec_full_data.csv")

# NSE bhavcopy URL pattern (equities)
# This pattern often works, but NSE may change it; keep configurable.
NSE_BHAV_URL <- function(d) {
  # d is Date; NSE bhavcopy usually: CM<DD><MMM><YYYY>BHAV.csv.zip
  # Example: https://archives.nseindia.com/content/historical/EQUITIES/2026/JAN/cm01JAN2026bhav.csv.zip
  yyyy <- format(d, "%Y")
  mon  <- toupper(format(d, "%b"))
  dd   <- format(d, "%d")
  sprintf(
    "https://archives.nseindia.com/content/historical/EQUITIES/%s/%s/cm%s%s%sbhav.csv.zip",
    yyyy, mon, dd, mon, yyyy
  )
}

# ---------------------------------------------------------------------------
# Core helpers
# ---------------------------------------------------------------------------

download_bhavcopy_for_date <- function(d, temp_dir) {
  url <- NSE_BHAV_URL(d)
  cat("Downloading bhavcopy for", as.character(d), "from:\n  ", url, "\n")
  
  resp <- tryCatch(
    {
      GET(url, add_headers(`User-Agent` = "Mozilla/5.0"))
    },
    error = function(e) {
      cat("  ❌ HTTP error:", e$message, "\n")
      return(NULL)
    }
  )
  
  if (is.null(resp) || http_error(resp)) {
    cat("  ❌ Failed (HTTP status:", if (!is.null(resp)) status_code(resp) else "NA", ")\n")
    return(NULL)
  }
  
  zip_path <- file.path(temp_dir, paste0("bhav_", format(d, "%Y%m%d"), ".zip"))
  writeBin(content(resp, "raw"), zip_path)
  
  # Unzip and read CSV
  files <- utils::unzip(zip_path, exdir = temp_dir)
  csv_file <- files[grep("\\.csv$", files, ignore.case = TRUE)][1]
  if (is.na(csv_file)) {
    cat("  ❌ No CSV in bhavcopy ZIP for", as.character(d), "\n")
    return(NULL)
  }
  
  cat("  ✓ Downloaded and extracted:", basename(csv_file), "\n")
  
  # Bhavcopy columns: SYMBOL, SERIES, OPEN, HIGH, LOW, CLOSE, LAST, PREVCLOSE,
  # TOTTRDQTY, TOTTRDVAL, TIMESTAMP, TOTALTRADES, ISIN
  df <- tryCatch(
    {
      read_csv(csv_file, show_col_types = FALSE)
    },
    error = function(e) {
      cat("  ❌ Error reading CSV:", e$message, "\n")
      return(NULL)
    }
  )
  
  if (is.null(df) || !"SYMBOL" %in% names(df) || !"TIMESTAMP" %in% names(df)) {
    cat("  ❌ Unexpected CSV structure for", as.character(d), "\n")
    return(NULL)
  }
  
  df <- df %>%
    mutate(
      TIMESTAMP = as.Date(TIMESTAMP, format = "%d-%b-%Y")
    ) %>%
    filter(!is.na(SYMBOL), SYMBOL != "", !is.na(TIMESTAMP))
  
  cat("  ✓ Rows loaded:", nrow(df), "\n")
  df
}

build_or_update_historical_file <- function(start_date, end_date) {
  if (start_date > end_date) {
    stop("start_date must be <= end_date")
  }
  
  cat("=== BUILDING / UPDATING NSE HISTORICAL SECURITIES FILE ===\n")
  cat("Project root :", PROJECT_ROOT, "\n")
  cat("Data dir     :", DATA_DIR, "\n")
  cat("Output file  :", HIST_FILE, "\n")
  cat("Date range   :", as.character(start_date), "to", as.character(end_date), "\n\n")
  
  existing <- NULL
  if (file.exists(HIST_FILE)) {
    cat("Loading existing historical file...\n")
    existing <- read_csv(HIST_FILE, show_col_types = FALSE)
    if (!"TIMESTAMP" %in% names(existing)) {
      stop("Existing nse_sec_full_data.csv has no TIMESTAMP column.")
    }
    existing <- existing %>%
      mutate(TIMESTAMP = as.Date(TIMESTAMP))
    cat("  ✓ Existing rows:", nrow(existing), "\n")
  }
  
  all_dates <- seq.Date(start_date, end_date, by = "day")
  temp_dir <- tempdir()
  
  new_data_list <- list()
  for (d in all_dates) {
    # Skip non-trading days quickly if already in existing
    if (!is.null(existing) && any(existing$TIMESTAMP == d)) {
      cat("Skipping", as.character(d), "- already in existing file\n")
      next
    }
    
    day_df <- download_bhavcopy_for_date(d, temp_dir)
    if (!is.null(day_df) && nrow(day_df) > 0) {
      new_data_list[[as.character(d)]] <- day_df
    }
  }
  
  if (length(new_data_list) == 0 && !is.null(existing)) {
    cat("\nNo new data downloaded; existing file is already up to date for this range.\n")
    return(invisible(existing))
  }
  
  new_data <- if (length(new_data_list) > 0) bind_rows(new_data_list) else NULL
  
  combined <- if (!is.null(existing) && !is.null(new_data)) {
    bind_rows(existing, new_data)
  } else if (!is.null(existing)) {
    existing
  } else {
    new_data
  }
  
  if (is.null(combined) || nrow(combined) == 0) {
    stop("No data available to write to nse_sec_full_data.csv")
  }
  
  # Deduplicate (keep latest TOTTRDVAL per SYMBOL/TIMESTAMP)
  combined <- combined %>%
    arrange(SYMBOL, TIMESTAMP, desc(TOTTRDVAL)) %>%
    distinct(SYMBOL, TIMESTAMP, .keep_all = TRUE) %>%
    arrange(TIMESTAMP, SYMBOL)
  
  write_csv(combined, HIST_FILE)
  
  cat("\n=== SUMMARY ===\n")
  cat("Total rows written :", nrow(combined), "\n")
  cat("Date range         :", as.character(min(combined$TIMESTAMP)), "to", as.character(max(combined$TIMESTAMP)), "\n")
  cat("Unique symbols     :", length(unique(combined$SYMBOL)), "\n")
  cat("File               :", HIST_FILE, "\n")
  cat("✅ Historical NSE securities file is ready.\n")
  
  invisible(combined)
}

# ---------------------------------------------------------------------------
# Main (when run as a script)
# ---------------------------------------------------------------------------

if (sys.nframe() == 0) {
  # Default: last 60 calendar days; adjust as needed
  end_date_default <- Sys.Date()
  start_date_default <- end_date_default - 60
  
  cat("Running download_nse_historical_data.R as a script.\n")
  cat("Default range: last 60 days (", as.character(start_date_default), "to", as.character(end_date_default), ")\n\n")
  
  build_or_update_historical_file(start_date_default, end_date_default)
}

File: python/core/download_nse_historical_data.py
#!/usr/bin/env python3
"""
Download and build historical NSE securities file: data/nse_sec_full_data.csv

This script:
- Downloads daily NSE equity bhavcopy files from the NSE archives
- Extracts and normalizes the data
- Builds or updates a consolidated historical CSV compatible with:
  - load_latest_data.R  (expects data/nse_sec_full_data.csv)
  - fixed_nse_universe_analysis.py
"""

from __future__ import annotations

import argparse
import io
import sys
from datetime import date, datetime, timedelta
from pathlib import Path
from typing import Optional

import pandas as pd
import requests
from zipfile import ZipFile, BadZipFile


# ---------------------------------------------------------------------------
# Configuration helpers
# ---------------------------------------------------------------------------

def get_project_root() -> Path:
    """
    Repo root: python/core/ → go up twice (core, python) to .../python, once more to repo.
    Writes consolidated CSV to <repo>/data/nse_sec_full_data.csv.
    """
    return Path(__file__).resolve().parent.parent.parent


PROJECT_ROOT: Path = get_project_root()
DATA_DIR: Path = PROJECT_ROOT / "data"
DATA_DIR.mkdir(exist_ok=True, parents=True)

HIST_FILE: Path = DATA_DIR / "nse_sec_full_data.csv"


def nse_bhav_url(d: date) -> str:
    """
    Construct the NSE bhavcopy URL for a given date.

    NSE equity bhavcopy archives typically follow the pattern:
    https://archives.nseindia.com/content/historical/EQUITIES/YYYY/MON/cmDDMONYYYYbhav.csv.zip
    """
    yyyy = d.strftime("%Y")
    mon = d.strftime("%b").upper()
    dd = d.strftime("%d")
    return (
        f"https://archives.nseindia.com/content/historical/EQUITIES/"
        f"{yyyy}/{mon}/cm{dd}{mon}{yyyy}bhav.csv.zip"
    )


# ---------------------------------------------------------------------------
# Core download & parsing logic
# ---------------------------------------------------------------------------

def download_bhavcopy_for_date(d: date, session: Optional[requests.Session] = None) -> Optional[pd.DataFrame]:
    """
    Download and parse the NSE bhavcopy for a single date.

    Returns:
        pandas.DataFrame with at least SYMBOL and TIMESTAMP columns,
        or None if download/parse fails (e.g., holiday, missing file).
    """
    url = nse_bhav_url(d)
    print(f"Downloading bhavcopy for {d.isoformat()} from {url}")

    sess = session or requests.Session()
    try:
        resp = sess.get(
            url,
            headers={
                # Use a browser-like User-Agent to avoid simple blocks
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)"
            },
            timeout=30,
        )
    except requests.RequestException as exc:
        print(f"  ❌ HTTP error for {d}: {exc}")
        return None

    if resp.status_code != 200:
        print(f"  ❌ HTTP {resp.status_code} for {d} (likely holiday or missing file)")
        return None

    try:
        with ZipFile(io.BytesIO(resp.content)) as zf:
            # Find first CSV file inside the ZIP
            csv_name = next((n for n in zf.namelist() if n.lower().endswith(".csv")), None)
            if not csv_name:
                print(f"  ❌ No CSV found in ZIP for {d}")
                return None

            with zf.open(csv_name) as f:
                df = pd.read_csv(f)
    except BadZipFile as exc:
        print(f"  ❌ Invalid ZIP for {d}: {exc}")
        return None
    except Exception as exc:
        print(f"  ❌ Error reading CSV for {d}: {exc}")
        return None

    # Basic structural validation
    required_cols = {"SYMBOL", "TIMESTAMP"}
    if not required_cols.issubset(df.columns):
        print(f"  ❌ Missing required columns in bhavcopy for {d}: {required_cols - set(df.columns)}")
        return None

    # Normalize and filter rows
    try:
        df["TIMESTAMP"] = pd.to_datetime(df["TIMESTAMP"], format="%d-%b-%Y").dt.date
    except Exception:
        # Fallback: let pandas infer date format if explicit parsing fails
        df["TIMESTAMP"] = pd.to_datetime(df["TIMESTAMP"], errors="coerce").dt.date

    df = df[
        df["SYMBOL"].notna()
        & (df["SYMBOL"] != "")
        & df["TIMESTAMP"].notna()
    ].copy()

    print(f"  ✓ Loaded {len(df)} rows for {d}")
    return df


def build_or_update_historical_file(start_date: date, end_date: date) -> pd.DataFrame:
    """
    Build or update the consolidated historical securities file.

    - Reads existing data from HIST_FILE if present
    - Downloads bhavcopies for the requested date range
    - Appends new rows and deduplicates by (SYMBOL, TIMESTAMP)
    - Writes the result back to HIST_FILE
    """
    if start_date > end_date:
        raise ValueError("start_date must be <= end_date")

    print("=== BUILDING / UPDATING NSE HISTORICAL SECURITIES FILE ===")
    print(f"Project root : {PROJECT_ROOT}")
    print(f"Data dir     : {DATA_DIR}")
    print(f"Output file  : {HIST_FILE}")
    print(f"Date range   : {start_date.isoformat()} to {end_date.isoformat()}\n")

    existing: Optional[pd.DataFrame] = None
    if HIST_FILE.exists():
        print("Loading existing historical file...")
        existing = pd.read_csv(HIST_FILE, low_memory=False)
        if "TIMESTAMP" not in existing.columns:
            raise ValueError("Existing nse_sec_full_data.csv has no TIMESTAMP column.")
        existing["TIMESTAMP"] = pd.to_datetime(existing["TIMESTAMP"]).dt.date
        print(f"  ✓ Existing rows: {len(existing)}")

    all_dates = [start_date + timedelta(days=i) for i in range((end_date - start_date).days + 1)]
    session = requests.Session()

    new_frames: list[pd.DataFrame] = []
    for d in all_dates:
        # Skip days already present in existing data, if available
        if existing is not None and (existing["TIMESTAMP"] == d).any():
            print(f"Skipping {d.isoformat()} - already present in existing file")
            continue

        day_df = download_bhavcopy_for_date(d, session=session)
        if day_df is not None and not day_df.empty:
            new_frames.append(day_df)

    if not new_frames and existing is not None:
        print("\nNo new data downloaded; existing file already covers this range.")
        return existing

    new_data = pd.concat(new_frames, ignore_index=True) if new_frames else None

    if existing is not None and new_data is not None:
        combined = pd.concat([existing, new_data], ignore_index=True)
    elif existing is not None:
        combined = existing
    elif new_data is not None:
        combined = new_data
    else:
        raise RuntimeError("No data available to write to nse_sec_full_data.csv")

    # Ensure key numeric fields are numeric to avoid later issues
    for col in ["OPEN", "HIGH", "LOW", "CLOSE", "TOTTRDQTY", "TOTTRDVAL"]:
        if col in combined.columns:
            combined[col] = pd.to_numeric(combined[col], errors="coerce")

    # Deduplicate by SYMBOL/TIMESTAMP, keeping row with highest TOTTRDVAL
    if "TOTTRDVAL" in combined.columns:
        combined = (
            combined.sort_values(["SYMBOL", "TIMESTAMP", "TOTTRDVAL"], ascending=[True, True, False])
            .drop_duplicates(subset=["SYMBOL", "TIMESTAMP"], keep="first")
        )
    else:
        combined = (
            combined.sort_values(["SYMBOL", "TIMESTAMP"])
            .drop_duplicates(subset=["SYMBOL", "TIMESTAMP"], keep="first")
        )

    combined = combined.sort_values(["TIMESTAMP", "SYMBOL"])

    # Write final consolidated CSV
    combined.to_csv(HIST_FILE, index=False)

    print("\n=== SUMMARY ===")
    print(f"Total rows written : {len(combined)}")
    print(f"Date range         : {combined['TIMESTAMP'].min()} to {combined['TIMESTAMP'].max()}")
    print(f"Unique symbols     : {combined['SYMBOL'].nunique()}")
    print(f"File               : {HIST_FILE}")
    print("✅ Historical NSE securities file is ready.")

    return combined


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------

def parse_args(argv: Optional[list[str]] = None) -> argparse.Namespace:
    """
    Parse command-line arguments for controlling the download range.
    """
    parser = argparse.ArgumentParser(
        description="Download NSE historical equity data and build data/nse_sec_full_data.csv"
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument(
        "--days",
        type=int,
        default=60,
        help="Number of most recent calendar days to download (default: 60)",
    )
    group.add_argument(
        "--start-date",
        type=str,
        help="Start date in YYYY-MM-DD format (overrides --days when used with --end-date)",
    )
    parser.add_argument(
        "--end-date",
        type=str,
        help="End date in YYYY-MM-DD format (used with --start-date; defaults to today)",
    )
    return parser.parse_args(argv)


def main(argv: Optional[list[str]] = None) -> None:
    """
    Main entry point: determine date range, then build/update the historical file.
    """
    args = parse_args(argv)

    today = date.today()

    if args.start_date:
        try:
            start = datetime.strptime(args.start_date, "%Y-%m-%d").date()
        except ValueError:
            raise SystemExit("Invalid --start-date, expected YYYY-MM-DD")
        if args.end_date:
            try:
                end = datetime.strptime(args.end_date, "%Y-%m-%d").date()
            except ValueError:
                raise SystemExit("Invalid --end-date, expected YYYY-MM-DD")
        else:
            end = today
    else:
        # Use relative range based on --days (optionally anchor --end-date for backfills)
        if args.days <= 0:
            raise SystemExit("--days must be a positive integer")
        if args.end_date:
            try:
                end = datetime.strptime(args.end_date, "%Y-%m-%d").date()
            except ValueError:
                raise SystemExit("Invalid --end-date, expected YYYY-MM-DD")
        else:
            end = today
        start = end - timedelta(days=args.days - 1)

    build_or_update_historical_file(start, end)


if __name__ == "__main__":
    main(sys.argv[1:])

